tests/ext/drv/test_ext.c

```diff
@@ -0,0 +1,489 @@
+ *
+ * @param client_binding_context Provider module's context for binding with the client.
+ * @retval STATUS_SUCCESS The operation succeeded.
+ * @retval STATUS_INVALID_PARAMETER One or more parameters are invalid.
+ */
+NTSTATUS
+_test_ebpf_extension_hook_provider_detach_client(void* provider_binding_context)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+
+    test_ebpf_extension_hook_client_t* local_client_context =
+        (test_ebpf_extension_hook_client_t*)provider_binding_context;
+    test_ebpf_extension_hook_provider_t* provider_context = NULL;
+
+    if (local_client_context == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    provider_context = &_test_ebpf_extension_hook_provider_context;
+    provider_context->attached_client = NULL;
+
+    ebpf_free(local_client_context);
+
+Exit:
+    return status;
+}
+
+void
+test_ebpf_extension_hook_provider_unregister()
+{
+    test_ebpf_extension_hook_provider_t* provider_context = &_test_ebpf_extension_hook_provider_context;
+
+    NTSTATUS status = NmrDeregisterProvider(provider_context->nmr_provider_handle);
+    if (status == STATUS_PENDING)
+        // Wait for clients to detach.
+        NmrWaitForProviderDeregisterComplete(provider_context->nmr_provider_handle);
+}
+
+NTSTATUS
+test_ebpf_extension_hook_provider_register()
+{
+    test_ebpf_extension_hook_provider_t* local_provider_context;
+    NTSTATUS status = STATUS_SUCCESS;
+
+    local_provider_context = &_test_ebpf_extension_hook_provider_context;
+
+    status = NmrRegisterProvider(
+        &_test_ebpf_extension_hook_provider_characteristics,
+        local_provider_context,
+        &local_provider_context->nmr_provider_handle);
+    if (!NT_SUCCESS(status))
+        goto Exit;
+
+Exit:
+    if (!NT_SUCCESS(status))
+        test_ebpf_extension_hook_provider_unregister();
+
+    return status;
+}
+
+ebpf_result_t
+test_ebpf_extension_invoke_program(_In_ test_program_context_t* context, _Out_ uint32_t* result)
+{
+    test_ebpf_extension_hook_provider_t* hook_provider_context = &_test_ebpf_extension_hook_provider_context;
+
+    test_ebpf_extension_hook_client_t* hook_client = hook_provider_context->attached_client;
+
+    ebpf_invoke_program_function_t invoke_program = hook_client->invoke_program;
+    void* client_binding_context = hook_client->client_binding_context;
+
+    // Run the eBPF program using cached copies of invoke_program and client_binding_context.
+    return invoke_program(client_binding_context, context, result);
+}
+
+// Helper Function Definitions.
+
+static int
+_test_ebpf_extension_helper_func1(test_program_context_t* context)
+{
+    UNREFERENCED_PARAMETER(context);
+    return 0;
+}
+
+static void
+_test_ebpf_extension_helper_func2(void* mem_pointer, uint32_t size)
+{
+    UNREFERENCED_PARAMETER(mem_pointer);
+    UNREFERENCED_PARAMETER(size);
+}
+
+static void
+_test_ebpf_extension_helper_func3(uint8_t param)
```

Comment:
avoid abbreviations.  (but `arg` is ok because C uses it as standard)
```suggestion
_test_ebpf_extension_helper_func3(uint8_t arg)
```

Reply:
renamed to arg.